patches/server/0099-Implement-EntityPickupItemEvent.patch

```diff
@@ -0,0 +1,68 @@
++                // event.setCancelled(!entityhuman.canPickUpLoot); TODO
++                this.world.getServer().getPluginManager().callEvent(entityEvent);
++
++                if (entityEvent.isCancelled()) {
++                    return;
++                }
++                // PandaSpigot end
+             }
+             // CraftBukkit end
+             boolean flag = this.fromPlayer == 1 || this.fromPlayer == 2 && entityhuman.abilities.canInstantlyBuild;
+diff --git a/src/main/java/net/minecraft/server/EntityInsentient.java b/src/main/java/net/minecraft/server/EntityInsentient.java
+index c1f100c3bd68329b77f9a899cc7353374064daf6..0b6427cc5d0bbb81217f3da261c2e0f4a4fdca3a 100644
+--- a/src/main/java/net/minecraft/server/EntityInsentient.java
++++ b/src/main/java/net/minecraft/server/EntityInsentient.java
+@@ -396,7 +396,14 @@ public abstract class EntityInsentient extends EntityLiving {
+                 }
+             }
+ 
+-            if (flag && this.a(itemstack)) {
++            // PandaSpigot start - EntityPickupItemEvent
++            boolean canPickup = flag && this.a(itemstack);
++            org.bukkit.event.entity.EntityPickupItemEvent entityEvent = new org.bukkit.event.entity.EntityPickupItemEvent((org.bukkit.entity.LivingEntity) this.getBukkitEntity(), (org.bukkit.entity.Item) entityitem.getBukkitEntity(), 0);
++            entityEvent.setCancelled(!canPickup);
++            this.world.getServer().getPluginManager().callEvent(entityEvent);
++
++            if (!entityEvent.isCancelled()) {
++            // PandaSpigot end
+                 if (itemstack1 != null && this.random.nextFloat() - 0.1F < this.dropChances[i]) {
+                     this.a(itemstack1, 0.0F);
+                 }
+diff --git a/src/main/java/net/minecraft/server/EntityItem.java b/src/main/java/net/minecraft/server/EntityItem.java
+index 1f8b1c726ef7343ee7feb1e4ea4c1ef8cc453723..a1ac66251044a5691baa6d2c44abaec847e224df 100644
+--- a/src/main/java/net/minecraft/server/EntityItem.java
++++ b/src/main/java/net/minecraft/server/EntityItem.java
+@@ -317,6 +317,16 @@ public class EntityItem extends Entity {
+                 if (event.isCancelled()) {
+                     return;
+                 }
++                // PandaSpigot start - EntityPickupItemEvent
++                // Call newer event afterwards
++                org.bukkit.event.entity.EntityPickupItemEvent entityEvent = new org.bukkit.event.entity.EntityPickupItemEvent(entityhuman.getBukkitEntity(), (org.bukkit.entity.Item) this.getBukkitEntity(), remaining);
++                // event.setCancelled(!entityhuman.canPickUpLoot); TODO
++                this.world.getServer().getPluginManager().callEvent(entityEvent);
++                if (entityEvent.isCancelled()) {
++                    return;
++                }
++                itemstack.count = canHold + remaining;
```

Comment:
Is this right? Above, `itemstack.count` is set to `canHold` before calling PlayerPickupItemEvent, but is changed to `canHold + remaining` before EntityPickupItemEvent is called, and set again to this value afterwards.

Reply:
changed to only run that after both methods it was confusing cause I was looking at 2 diff sources